### ISSUE TITLE
ghc-7.10.1 compatibility

### ISSUE THREAD

**Issue by KblCb** [basvandijk/usb#10]
Hi!

I have tried to port usb to base-4.8.0.0 and have got a trable. New [registerFd](http://hackage.haskell.org/package/base-4.8.0.0/docs/GHC-Event.html#v:registerFd) takes extra Lifetime argument. Lifetime is ADT that does not exported from GHC.Event.Internal.

Is it possible to workaround using of registerFd?

**Comment by basvandijk** [comment]
Thanks for the report. @bgamari do you know where I should import Lifetime from?

**Comment by KblCb** [comment]
I have looked through ghc source code and think that `Lifetime` is nowhere imported. `Lifetime` is defined in GHC.Event.Internal as:
```haskell
data Lifetime = OneShot | MultiShot
```
Since `Lifetime` is isomorphic to `Bool`, i use `unsafeCoerce True` as a `Lifetime` value. It's ugly hack, but usb-example works.

**Comment by bgamari** [comment]
@basvandijk indeed this is a pretty awful oversight on my part. Sorry, I'll need to get a patch into 7.10.2 to fix this. Beyond that all I can recommend is the `unsafeCoerce` trick that @KblCb suggests. Ugh.

**Comment by KblCb** [comment]
@basvandijk Do you need a pool requset?

**Comment by basvandijk** [comment]
@bgamari no worries, it would be great if you can get a fix in 7.10.2 though.

@KblCb a pull request with the `unsafeCoerce` hack would be much appreciated. Please also make sure you test the patch with GHC 7.6 and 7.8. Thanks!

**Comment by bgamari** [comment]
@basvandijk I believe this should do the trick: https://phabricator.haskell.org/D838

**Comment by basvandijk** [comment]
@bgamari Thanks. Do you know the ETA of 7.10.2?

**Comment by basvandijk** [comment]
@bgamari it would be great if the base library could be bumped to 4.8.1.0 instead of 4.8.0.1. Then I can test for the presence of this version using

```cpp
#if MIN_VERSION_base(4,8,1)
```

**Comment by KblCb** [comment]
@basvandijk I got the hang of my code in 11438191e27538c1f2512f3f6ea55c4534874ded:
```haskell
action dev outEndpoint = withDeviceHandle dev $
    \hnd -> withDetachedKernelDriver hnd 0 $ withClaimedInterface hnd 0 $ do
        writeBulk hnd outEndpoint "C" 1000
        writeBulk hnd outEndpoint "C" 1000
```
But everything worked in faf4a108ddfceea65d18077b6fd4892e3351fa7d.

**Comment by basvandijk** [comment]
Mmm maybe I didn't get the multi-shot implementation right. I'll try to reproduce it on my system.

Thanks for testing!

**Comment by basvandijk** [comment]
I can confirm that `MultiShot` is not working properly.

I'm using  [usb-example](https://github.com/basvandijk/usb-example/blob/master/example.hs)  to debug this.

When I add a `traceShowM ("registerAndInsert", fd, evt)` in [registerAndInsert](https://github.com/basvandijk/usb/blob/11438191e27538c1f2512f3f6ea55c4534874ded/System/USB/Base.hs#L396) and a `traceShowM ("callback", _fdKey, _ev)` in [callback](https://github.com/basvandijk/usb/blob/11438191e27538c1f2512f3f6ea55c4534874ded/System/USB/Base.hs#L414) and run `example` on some attached USB device I get the following log (#annotated for explanation): 

```
sudo dist/build/example/example 0x80ee 0x0021

# Here you see some libusb control files being registered:
("registerAndInsert",13,1)
("registerAndInsert",15,1)
("registerAndInsert",17,1)

Waiting for device attachment...
[timestamp] [threadID] facility level [function call] <message>
--------------------------------------------------------------------------------
[ 0.003607] [0000316b] libusb: debug [libusb_get_device_list] 
deviceSpeed:   FullSpeed
busNumber:     2
portNumber:    1
portNumbers:   [1]
deviceAddress: 2

Opening device...
[ 0.003986] [0000316b] libusb: debug [libusb_open] open 2.2
[ 0.004034] [0000316b] libusb: debug [usbi_add_pollfd] add fd 18 events 4

# Opening a device causes another file to be registered.
("registerAndInsert",18,4)

# And some event on 13 took place because its callback is being called:
("callback",FdKey {keyFd = 13, keyUnique = 1},[evtRead])

[ 0.004342] [0000316c] libusb: debug [libusb_try_lock_events] someone else is modifying poll fds
[ 0.004352] [0000316c] libusb: debug [libusb_event_handler_active] someone else is modifying poll fds
[ 0.004357] [0000316c] libusb: debug [libusb_handle_events_timeout_completed] another thread is doing event handling
Detaching kernel driver...
[ 0.004498] [0000316b] libusb: debug [libusb_kernel_driver_active] interface 0
[ 0.004550] [0000316b] libusb: debug [libusb_detach_kernel_driver] interface 0
Claiming interface...
[ 0.037799] [0000316b] libusb: debug [libusb_claim_interface] interface 0
[ 0.037944] [0000316b] libusb: debug [libusb_get_config_descriptor] index 0
maxPacketSize = 8
Creating transfer...
(1/3) reading 8 bytes during a maximum of 5000 ms...
[ 0.038534] [0000316b] libusb: debug [add_to_flying_list] arm timerfd for timeout in 5000ms (first in line)
[ 0.038650] [0000316b] libusb: debug [submit_bulk_transfer] need 1 urbs for new transfer with length 8

# Because my device doesn't return data I should get a timeout event after 5000ms. This is correctly happening here: 
("callback",FdKey {keyFd = 17, keyUnique = 3},[evtRead])

[ 5.039339] [0000316c] libusb: debug [libusb_handle_events_timeout_completed] doing our own event handling
[ 5.039375] [0000316c] libusb: debug [handle_events] poll() 4 fds with timeout in 0ms
[ 5.039395] [0000316c] libusb: debug [handle_events] poll() returned 1
[ 5.039407] [0000316c] libusb: debug [handle_events] timerfd triggered
[ 5.039429] [0000316c] libusb: debug [libusb_cancel_transfer] 
[ 5.058627] [0000316c] libusb: debug [disarm_timerfd] 
[ 5.058666] [0000316c] libusb: debug [handle_events] poll() 4 fds with timeout in 0ms
[ 5.058675] [0000316c] libusb: debug [handle_events] poll() returned 1
[ 5.058686] [0000316c] libusb: debug [reap_for_handle] urb type=1 status=-2 transferred=0
[ 5.058692] [0000316c] libusb: debug [handle_bulk_completion] handling completion status -2 of bulk urb 1/1
[ 5.058698] [0000316c] libusb: debug [handle_bulk_completion] abnormal reap: urb status -2
[ 5.058703] [0000316c] libusb: debug [handle_bulk_completion] abnormal reap: last URB handled, reporting
[ 5.058710] [0000316c] libusb: debug [usbi_handle_transfer_cancellation] detected timeout cancellation
[ 5.058716] [0000316c] libusb: debug [disarm_timerfd] 
[ 5.058724] [0000316c] libusb: debug [usbi_handle_transfer_completion] transfer 0x8653d0 has callback 0x7fd1e9e42098
Reading timed out!
Read 0 bytes:

# Lets do another read from the same device:

(2/3) reading 16 bytes during a maximum of 5000 ms...
[ 5.058971] [0000316b] libusb: debug [add_to_flying_list] arm timerfd for timeout in 5000ms (first in line)
[ 5.058991] [0000316b] libusb: debug [submit_bulk_transfer] need 1 urbs for new transfer with length 16

# We started reading again from the same device but this time no callback is being called which hints that the file has been unregistered with the IOManager.
```

When I change `usb` to use `OneShot` semantics the example works correctly.

@bgamari any idea what might be going wrong?

**Comment by bgamari** [comment]
@basvandijk hmm, let me have a closer look.

**Comment by bgamari** [comment]
@basvandijk I found the issue. Unfortunately it's a nasty bug which breaks multi-shot support in 7.10.1 quite badly.  :(

**Comment by bgamari** [comment]
@basvandijk I have opened GHC bug [#10317](https://ghc.haskell.org/trac/ghc/ticket/10317) to track this issue. I believe [D849](https://phabricator.haskell.org/D849) should fix this although this code is in terrible need of more tests.

**Comment by basvandijk** [comment]
I have released [usb-1.3.0.1](http://hackage.haskell.org/package/usb-1.3.0.1) which for GHC-.7.10.1 keeps using OneShot registrations. It switches to MultiShot when base >= 4.8.1.0 is being used. 

@KblCb and @bgamari thanks for your work on this!

@bgamari I'm looking forward to your fix in GHC-7.10.2 / base-4.8.1.0!
